### ISSUE TITLE
Reduce size overhead of corehost

### ISSUE THREAD

**Issue by Genbox** [dotnet/runtime#68593]
My company have a size restriction on our software, which is written entirely in .NET 6. We use singlefile as the primary means of publishing, but noticed that the size of the final package is rather large compared to our application code.

It turns out the singlefile host is optimized for speed rather than size. As an experiment, we changed the host's CMake files to optimize for size.

File1: https://github.com/dotnet/runtime/blob/main/src/native/corehost/CMakeLists.txt
Here we added `add_compile_options(-Os)`

File2: https://github.com/dotnet/runtime/blob/main/eng/native/configureoptimization.cmake
Here we changed `Ox` to `Os`

Docs on Ox: https://docs.microsoft.com/en-us/cpp/build/reference/ox-full-optimization?view=msvc-170

The result is an apphost that is several megabytes smaller. I believe we went from 6 MB overhead to around 3 MB overhead. A reduction of about 3 MB in the final singlefile executable.

Also note that enabling singlefile compression has no impact on the apphost overhead since the compressed blob is embedded next to the apphost.

I don't see any reason why the apphosts should be optimized for speed. After all, they have little to no operations that require optimized code, and in any case, a 50% overhead in assembly size will probably negate the optimizations. It should probably default to -Os instead of -Ox.

**Comment by vitek-karas** [comment]
The host in case of single-file (assuming it's self-contained publish) contains the entire runtime+jit+gc, so I would argue it needs to be as fast as possible. That said, maybe there's some middle ground, but it would have to be well tested/measured.

**Comment by AaronRobinsonMSFT** [comment]
@Genbox It sounds like you already have the assets needed to run a few benchmarks and collect some of that data. Would it be possible for you to use Visual Studio's CPU profiler or PerfView or some other tool and collect details on your scenario. Hearing about some initial findings might help to move this forward.

**Comment by Genbox** [comment]
I'll do a few things/tests and see what the differences are.

**Comment by deeprobin** [comment]
Couldn't we introduce the possibility to use a property in `.csproj` to determine the optimization level of the host?

**Comment by jkotas** [comment]
> Couldn't we introduce the possibility to use a property in .csproj to determine the optimization level of the host?

It would mean building, testing and shipping yet another build of coreclr, for all platforms and OSes. It is expensive.

There are number of other ways one may want to customize the single file host, e.g. using a custom compression - https://github.com/dotnet/runtime/issues/61073.

I think the best way to satisfy all these requests is to publish recipe for how to build a host from cloned dotnet/runtime repo sources, with any arbitrary customizations.

**Comment by deeprobin** [comment]
> It would mean building, testing and shipping yet another build of coreclr, for all platforms and OSes. It is expensive.

That's right. The build is quite expensive. I mean, we still have a lot of optimization potential in the build, but I should move that to another issue after I've analyzed it in more detail.

However, there are many users who would probably prefer this size-reduced build. The question is: The build is expensive, but does the benefit relativize this price?

I personally have a few tools at work where file size is critical (due to low data transfer rates), for which this version of the host would be very good.

I like to hear other viewpoints :D

> I think the best way to satisfy all these requests is to publish recipe for how to build a host from cloned dotnet/runtime repo sources, with any arbitrary customizations.

👍🏼

**Comment by agocke** [comment]
One question: are you using the self-contained single file host? Or the framework-dependent one?

And if you're using self-contained, are you also using IL trimming? I generally want to understand the full scenario. 

In most cases where people are using self-contained the size of the runtime is usually dwarfed by the size of the IL.

**Comment by deeprobin** [comment]
Actually in my case not (Framework Dependant). IL trimming will make a lot of difference, you're right.

**Comment by Genbox** [comment]
I've run a few tests on the binary size alone. Just to get a feeling of what can be gained.
It is an x64/release configuration. The executable tested is the full single-file host.

Format: <compiler options>:<size in bytes>

1. /Ox (baseline): 8.097.792
2. /O1: 6.819.328
3. /Oxs: 6.819.328
4. /Os /GF /Ob1: 6.693.888
5. /O2: 8.097.792

I'm not sure why it seems some compiler flags have no impact on size. I'm not familiar with the cmake compiler or the C++ build chain, so maybe I'm doing something wrong.

JIT/GC compiler flags are a separate discussion, so I guess the question is if it would be possible to distribute a speed-optimized version (/Ox or /O2) and a size-optimized version (/Oxs).

What we did is to compile it with /Oxs and use distribute the nuget package on a private feed. Then we hacked it into to the msbuild chain with this:

Note that $CI is a variable we set to true on our $CI build system and `Custom.Microsoft.NETCore.App.Host.*` is the name of our custom build nuget package. 

```
  <Target Name="SmallHost" BeforeTargets="ProcessFrameworkReferences" Condition="'$(CI)'=='true' AND $(RuntimeIdentifier.StartsWith('win-'))" >
    <Message Importance="High" Text="######## Using small singlefilehost ########"/>
    <ItemGroup>
      <KnownAppHostPack Update="@(KnownAppHostPack)" AppHostPackNamePattern="Custom.Microsoft.NETCore.App.Host.**RID**"/>
    </ItemGroup>
  </Target>
```

So if there are other plans to support a size-minimized build (such as for AWS Lambda functions, size-sensitive apps or embedded programming) then it could just be a small change to produce the size optimized host as a nuget package. Something like `Microsoft.NETCore.App.Host.Embedded.win-x64`.
